### DISCUSSION TITLE
Is it risky to consider marginal emissions only?

### DISCUSSION THREAD

**Discussion by andrew-woosnam** [Green-Software-Foundation/sci#351] [SCI feedback]
Hey folks! I'm fairly new to carbon measurement, so I've been reading up on the difference between "average" and "marginal" emissions lately. As I understand it, "average" measures the carbon intensity of the grid as a whole, whereas "marginal" is just the intensity of the current marginal power producer (i.e. the 1 plant that's not running at full capacity & is dynamically adjusting its output to match demand). 

I see why the marginal signal is so important when making temporal decisions about electricity consumption; extra demand will be served by the marginal producer, incurring emissions at the marginal rate. But if I'm understanding the SCI [docs](https://github.com/Green-Software-Foundation/software_carbon_intensity/blob/main/Software_Carbon_Intensity/Software_Carbon_Intensity_Specification.md) correctly, the marginal signal is the _only_ emissions factor that influences the final score:

> "The only figure that matters when trying to optimize the scheduling of a computation in real-time is the marginal emissions intensity. This is the emissions intensity of the marginal power plant which will need to be turned up if a computation is scheduled (e.g., increase electricity demand from the grid) at that moment."

While I don't think this is wrong, per se, I'm wondering if there's room for improvement. Has this group already discussed this point [raised](https://www.electricitymaps.com/blog/marginal-vs-average-real-time-decision-making) by ElectricityMaps Founder Olivier Corradi?

> [PARAPHRASE]: **By looking at marginal emissions _alone_ and not considering average emissions at all, we miss out on opportunities to incentivize grid infrastructure improvement via the construction of new renewable generators.**

Here’s the scenario he's worried about: 
- A new fleet of EVs needs to decide when to charge from a grid supplied by majority solar / marginal gas during the day, and full gas at night.
- Marginal data alone would suggest that it doesn’t matter when the EVs charge, because the marginal rate never changes over the course of the day (the marginal power producer is always gas). The marginal signal suggests that charging at night makes just as much sense as charging during the day.
- In the _short term_, it’s true that it wouldn’t matter where the extra load goes; gas burnt at night emits just as much as the same amount of gas burnt during the day.
- In the _long term_, however, daytime consumption would likely improve the grid’s infrastructure more than nighttime consumption would.
  - More daytime demand strongly incentivizes the construction of additional solar (cheapest option to invest in).
  - Since solar isn't available at night, the next-cheapest option to accommodate extra nighttime demand would likely be to construct additional fossil plants.
    - Storage (e.g. batteries, pumped hydro, etc.) could also be an option for meeting extra nighttime demand, but ideally all demand is served by recently-produced renewable energy. That is: using electricity from wind/solar as soon as it’s generated is more efficient (less transitory waste) than it would be to store that energy for later use, and doesn't incur as many embodied emissions as the construction of energy storage facilities.
- So if we consider what happens when electricity demand for electricity begins to regularly approach the upper limit grid's capacity & additional power producers are needed:
  - If this chronically increased load occurs during the daytime, the cheapest option for additional generators is solar.
  - If this chronically increased load occurs at night, then the next best option would probably be fossils (assuming wind is not reliable enough & storage is too expensive).

So even if the marginal emission rate stays the same all the time, there may still be “better” and “worse” times to add extra demand?

Olivier Corradi also includes in his article:
> [research is currently underway](https://www.nrel.gov/analysis/cambium.html) to assess whether long-term effects can be incorporated into a marginal signal by investigating how the future grid would change if demand systematically was allocated to specific times. However, at present, no method currently exists for estimating future structural changes in real-time.

He goes on to suggest that the average signal does a better job at incentivizing investment in low-carbon power generation:
> It turns out that on this simplified grid, the average signal, by incentivizing electricity use during sunny hours, incentivizes the proper long-term behavior. [The] average signal incorporates all power plants of the grid, and thus incentivizes using electricity during times where the grid is clean. Thus, it is a signal that motivates the alignment of consumption and low-carbon production.

Personally I'm not sure how to reconcile short-term improvement motivated by marginal signals with long-term effects that might lead to better outcomes based on following average signals. Clearly we need some of both in the world: avoid emissions wherever we can right now (i.e. develop consumption strategies based on marginal emissions) while still trying to shift grid infrastructure away from fossils & toward renewables (which could be achieved via consumption strategies that take average signals into account). 

### So I have some questions for you all!
1. Does this observation seem sensible to you? Or is it missing something I haven't considered yet?
2. Do you know of any current work related to incorporating "long term" effects into carbon intensity metrics?
3. The "EV fleet charging" scenario posed above is admittedly based on a hypothetical & simplified grid. I don't know enough about the real world to know the answer to this: is this even a realistic scenario? Is it true that the marginal signal doesn't change (much) over the course of the day for some grids?
4. When making decisions about consumption, it makes a lot of sense to me (and would hopefully not be too hard) to make a habit of double-checking average emissions in addition to just the current marginal rate, just to ensure that even if the marginal signals don't provide a clearly "optimal" consumption time, we're still trying to pick a time that's likely to be servable be renewables. Is this something we could advocate for in some way? E.g. could we make sure the Carbon Aware SDK exposes average signals in addition to marginal ones? Or write up some communication encouraging consumers to choose times when renewables are available whenever it makes sense to do so?
5. Is there more info on this topic someone could point me to? (I did some searching around this GitHub org but nothing jumped out right away).

Thanks for making it through this tome! Looking forward to hearing your thoughts :)

**Comment by Henry-WattTime** [comment]
Hello Andrew, this is an interesting set of questions. 

We're starting to explore what real-time signals might help drive future investment in renewable generation. We've hypothesized that signals like renewable curtailment, the difference in revenue to renewable and fossil generators, and the alignment with potential long run output of renewable generators might be effective signals. I've actually seen very little evidence that optimizing for average emissions does support the future development of renewables. For example, average emissions would not as effectively predict incremental revenue to renewable generators as a signal that factors in the actual price of energy at that time. 

The simplified grid example also obscures many dynamics of real grids that contain a mix of generating resources that might be marginal. In a grid that contains oil, natural gas, and coal generators, the marginal emissions will change throughout the day and you can take advantage of that difference with load flexibility. Furthermore, the simplified example doesn't account for the different efficiencies of different power plants of the same fuel type that are operating (and if you want to go deeper down that hole, a single generator operates at different efficiencies under different load and weather conditions).  

An average emission signal may also mislead you about the actual emissions effect you are having in the real world. For example, the Pacific northwest has a low average emissions rate (lots of hydro), but if you add load in that region, there is no more hydro available to serve that demand (we're not building more dams/water behind a dam is a finite resource), so any additional electricity demand has to be met with additional fossil generation. A [recent paper](https://pubs.acs.org/doi/abs/10.1021/acs.est.2c02344) from Eric Hittinger shows that an average emissions signal really does not predict real-world change in emissions well, but a short-run marginal signal does. (I find this [twitter thread]( https://twitter.com/ElephantEating/status/1590741977880211458) from him very helpful).

Is this a good starting point? Would you be interested in discussing this in more depth?

**Comment by Henry-WattTime** [comment]
WG: Will invite @atwoosnam to the next WG meeting to participate in discussion
Should establish regular discussion of topics
-Many participants in WG are software experts, not grid experts
-Bring whole group up in grid emissions knowledge

**Comment by ciril-emaps** [comment]
Hi Andrew, many thanks for opening this discussion point and thanks a lot Henry for adding your thoughts. I believe this is a crucial topic to cover and it is great that we are engaging in more discussion around this. 

Since the debate about which emission factor or other possible signal is the best for decision-making is still ongoing, it is risky to prescribe marginal emissions as the only valid signal in the SCI. A suggestion could be to allow both marginal and average emission signals, and potentially other signals such as renewable potential profiles. In the end, **the intention and goal are to ensure the decarbonisation of electricity systems and with it our economy.** 

Regarding the EV case provided by Olivier, while it is true that this is a simplified grid, the simplification was done to facilitate understanding. Please see below power system data from CAISO that would be representative for this case. As you can see, the marginal signal stays relatively constant, while the average signal (similar to renewable potential signals) reflect solar generation. According to the optimal marginal signal, your best action is to move your load to late hours where very little solar is produced. If fleet-wide load-shifting is conducted like this on a regular basis, there will be no incentive to expand solar generation. This is not a unique occurrence and If needed, we can provide more such curves for other grid regions. 

<img src="https://user-images.githubusercontent.com/113919390/218122882-65713fdf-fbcc-47fe-81b8-6aa861246cdb.png" width = "500" height = "550" />

The [paper](https://pubs.acs.org/doi/abs/10.1021/acs.est.2c02344) highlighted by Henry is very insightful, but it seems that it focuses on emissions reduction from the addition of new technologies to the grid. This use case is very different from the load-shifting use case where you will not increase overall electricity demand, but instead change the demand curve to meet cleaner production. As the paper states, the analysis was conducted on a modeled grid and thus their result cannot be applied to real grids without further investigation. There is a great [Twitter thread](https://twitter.com/AukeHoekstra/status/1620525739803291648) directly responding to the Twitter thread Henry posted, which highlights the issues of marginal emission factors. 

Regarding research, Pieter Gagnon (NREL) has published a [paper](https://www.cell.com/iscience/fulltext/S2589-0042(22)00185-7?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2589004222001857%3Fshowall%3Dtrue) comparing different emission factors and recently published a [letter](https://www.pnas.org/doi/10.1073/pnas.2211624119) highlighting the shortcomings of short-term marginal emission factors to account for the long-term impact of grid interventions. Only long-run marginal emission factors are able to assess that, but these factors are currently not available at the required temporal and spatial granularity. Whether average signals, renewable potential profiles or other signals are better signals for decision-making is yet unclear, what is clear however is that short-run marginal is not suitable for long-term impact assessment.

As you can see, there is still a big need for further discussions and it very much benefits not only the SCI but the wider society to engage in further conversations.

**Comment by andrew-woosnam** [comment]
Hey folks, thanks for weighing in! Really appreciate your responses -- here are some thoughts you catalyzed:

@Henry-WattTime 
> I've actually seen very little evidence that optimizing for average emissions does support the future development of renewables.

I buy this. It makes sense to me that decisions to construct new power generation facilities are motivated by sundry factors (probably not all of which are totally in sync with scientific observations). For that reason I totally believe it's not easy to confidently draw cause -> effect relationships in this context. 

Having said that, I'm still persuaded by @ciril-emaps 's example illustrated in those graphs. I.e. it makes sense to me that instead of just "always opt to consume when marginal emissions are lowest", we might want to consider something like this:
- _opt for daytime consumption, _unless_ it would incur significantly* more emissions to add my load at night_
- i.e. aim to consume when the marginal emissions rate is lowest, _unless_ that marginal minimum occurs when average emissions are high AND that marginal minimum is very similar* to the marginal rate when the average signal is low.
- ( * this is admittedly subjective -- idk how this value would be decided)

Note I'm not advocating for an _abandonment_ of marginal -- Henry's Pacific Northwest example makes it clear that we shouldn't rely on average signals ALONE. But it still seems to me that even in the hydro-dominant Pacific Northwest it would be a wasted opportunity to schedule EV charging at night when you could otherwise do so via solar (if not today, then likely soon -- plenty of solar arrays are currently bogged down in the process of getting properly connected to the grid, no?)

Of course neither extreme would be ideal. We definitely wouldn't want to say "everyone consume 100% of your demand during the day (when average signal is low)" if -- like in PNW -- there's not enough renewable power for that much load AND the marginal supplier is very dirty (that sounds counterproductive; we wouldn't want to adopt behaviors that incur SO many extra emissions just in the hopes that it'll incentivize the construction of cleaner plants -- we're still trying to avoid as much atmospheric CO2 as possible). But at the same time, it still feels somewhat shortsighted to suggest adopting any behavior that would result in scheduling lots of extra load during the night in any place where the sun shines during the day.

Again, this only makes sense to me for those cases (and maybe there aren't many of them!) where the marginal signal is (effectively) constant. I.e. we shouldn't trust average signals at the _expense_ of the marginal (we still want to minimize actual emissions wherever we can), but surely there are cases where we have equal marginal values at different times. In those cases, shouldn't we always go for the option with the highest potential to be served by renewables? (In my mind this just means "daytime" -- maybe that's distinct from the average signal?)

Realistically though, with real-life marginal measurements I bet it's rarely possible to identify periods where the marginal signal is truly "constant". I'm guessing that it's almost always the case that the same load consumed in two different time periods A & B will "cause" (marginal) _some_ differing amount of emissions _e_ where _e > 0_. In the sense that we want to prevent emissions wherever/whenever we can, we'd say that we should always pick scenario A if it emits _any_ fewer emissions than scenario B, right? That sounds to me like an argument for why average signals shouldn't matter at all; always just opt for the lowest marginal signal because we need less CO2 ASAP.

But I'm still nervous about how marginal-only considerations might lead to counterintuitive outcomes like Ciril's EV fleet example. 

It feels to me like by ignoring average signals and using marginal alone, we're strictly reporting on a comparatively microscopic view of the grid: the precise carbon intensity of the grid at this precise moment in time. This doesn't sound like a bad thing -- if more solar plants come online, the marginal signal should shift accordingly to reflect the increased capacity able to be served by renewables, right? In which case we're fine to continue looking at marginal-alone for decision making, because we're optimizing for minimal carbon at every step of the way. 

This feels like a reactionary way to make decisions around carbon intensity. That's not meant to sound pejorative, just to observe that IF infrastructural changes were to happen to the grid, then the model might start reporting new values for carbon intensity (i.e. _after_ the fact, not in any sort of recommendatory way _before_ the change). Clearly there's great utility in a model like this -- I think we're all in agreement there that we need a practical way to answer "what emissions will my load incur if I start consuming right now"?

When it comes to decision making though, I still think there's something missing if folks using a model like this conclude that it's good to schedule large-scale consumption during times that'll be much harder to serve with renewables than alternative times. I.e. even if it saves more carbon right now, is it a good long term decision? Shouldn't the end goal be to supply as much demand as possible with "free" solar power?

So maybe this is more of a discussion about whether or not it's appropriate/beneficial for a model like SCI to make considerations about the future of the grid / where we'd like to shift consumption, or if it should strictly be a reporting mechanism that reflects only the here-and-now. It's definitely useful to know the answer to the here-and-now, but should there be an added layer of consideration before we use this information to determine our consumption behaviors?

(Maybe this analogy could work: by looking at marginal signals only, are we following a [Greedy](https://en.wikipedia.org/wiki/Greedy_algorithm) approach to decarbonization? Is that ok? I truly don't have an opinion here yet.)

I wish this could've been less of a ramble and more of an essay, but I don't have a thesis! 😄 
Any thoughts from you all? Would love to hear more opinions / info I haven't considered.

**TL;DR**
- Marginal is crucial; average is not a replacement for marginal.
- Folks can still plausibly commit to "bad" consumption behavior in the long-term by looking at marginal alone.
  - example = EV fleet charging at night
- If marginal is otherwise constant, shouldn't we opt to consume during periods of lower average emissions?
- Can we help incentivize "good" long-term consumption schedules somehow?
  - Is the average signal an appropriate way to do this? Or could it be simplified to just "sunlight hours"?
- Is marginal ever truly "constant"? Between options A & B, one's marginal emissions must almost certainly be lower by _some_ amount ... shouldn't we always opt to consume during the option with lowest marginal emissions?
  - On one hand, yes because we want to avoid emissions wherever possible 
  - On the other hand, isn't the end goal to shift consumption to times that could be most-easily served by renewables? (e.g. sunlight hours)
- By looking at marginal signals only, are we following a [Greedy](https://en.wikipedia.org/wiki/Greedy_algorithm) approach to decarbonization? Is that ok?

**Comment by liamlaverty** [comment]
Hey, this thread is great, and I think a lot of people end up here with similar questions & thoughts to @andrew-woosnam , as it's linked to from some prominent blogs on the subject of decarbonising compute.

Is there an update related to the discussion here? Especially interested to know if this suggestion from @ciril-emaps to provide both marginal and avg emissions signals was implemented?

>A suggestion could be to allow both marginal and average emission signals, and potentially other signals such as renewable potential profiles. In the end, the intention and goal are to ensure the decarbonisation of electricity systems and with it our economy.

**Reply by mrchrisadams** [reply]
Hi @liamlaverty, this was merged with PR #353 

https://github.com/Green-Software-Foundation/sci/pull/353
